File: fetch-changelog.py
import requests
import re
import os
import sys
import time
import csv
from dotenv import load_dotenv
from html import unescape

# Load environment variables from .env file
load_dotenv()

GITHUB_TOKEN = os.getenv("GITHUB_TOKEN")
REPO_OWNER = "woocommerce"
REPO_NAME = "woocommerce"

def fetch_changelog(version):
    # Use the trunk changelog URL
    changelog_url = "https://raw.githubusercontent.com/woocommerce/woocommerce/refs/heads/trunk/changelog.txt"

    try:
        # Fetch the changelog
        response = requests.get(changelog_url)
        response.raise_for_status()  # Raise an exception for bad status codes
        
        # Get the content and split at the changelog marker
        content = response.text
        
        # Find the specific version section
        version_marker = f"= {version} "
        if version_marker not in content:
            print(f"Changelog section for version {version} not found. Attempting to generate from GitHub API...")
            return fetch_prs_from_github(version)
            
        # Split at the version marker and get the content up to the next version
        parts = content.split(version_marker)
        if len(parts) < 2:
            print(f"Error: Could not find changelog section for version {version}")
            return False
            
        # Get the content up to the next version marker
        changelog_content = parts[1].split("= ")[0].strip()
        
        # Save the changelog to a file
        save_changelog(version, changelog_content)
        return True

    except requests.exceptions.RequestException as e:
        print(f"Error fetching changelog: {e}")
        return False
    except Exception as e:
        print(f"An error occurred: {e}")
        return False

def fetch_prs_from_github(version):
    """Fetch PRs from GitHub API for the specified milestone. If test_mode, only fetch 10 PRs."""
    if not GITHUB_TOKEN:
        print("Error: GITHUB_TOKEN environment variable is not set")
        return False
        
    # First verify the milestone exists
    milestone_url = f"https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}/milestones"
    headers = {
        "Authorization": f"token {GITHUB_TOKEN}",
        "Accept": "application/vnd.github.v3+json"
    }
    
    try:
        # Get all milestones to find the correct one
        response = requests.get(milestone_url, headers=headers)
        if response.status_code == 403:
            print("Error: GitHub API rate limit exceeded. Please try again later.")
            return False
        response.raise_for_status()
        milestones = response.json()
        
        # Find the milestone that matches our version
        milestone_number = None
        for milestone in milestones:
            if milestone['title'] == version:
                milestone_number = milestone['number']
                print(f"Found milestone {version} with number {milestone_number}")
                break
        
        if not milestone_number:
            print(f"Warning: Could not find milestone for version {version}")
            print("Available milestones:")
            for m in milestones:
                print(f"- {m['title']}")
            return False

        # Now fetch PRs with the correct milestone number
        url = f"https://api.github.com/repos/{REPO_OWNER}/{REPO_NAME}/issues"  # Use issues endpoint instead of pulls
        params = {
            "state": "closed",
            "milestone": str(milestone_number),
            "per_page": 100
        }

        all_prs = []
        next_url = url
        page_count = 0

        while next_url:
            page_count += 1
            print(f"Fetching PRs from page {page_count}...")
            response = requests.get(next_url, headers=headers, params=params)
            response.raise_for_status()
            issues = response.json()

            if not issues:
                print("No more PRs found, stopping pagination.")
                break

            # Filter to only include PRs (issues with pull_request field)
            prs = [issue for issue in issues if 'pull_request' in issue]
            print(f"Found {len(prs)} PRs on this page")
            all_prs.extend(prs)

            # Get the next page URL from the Link header
            link_header = response.headers.get('Link', '')
            if link_header:
                # Parse the Link header to find the 'next' URL
                for link in link_header.split(','):
                    if 'rel="next"' in link:
                        next_url = link.split(';')[0].strip('<>')
                        break
                else:
                    next_url = None
            else:
                next_url = None
                
            if not next_url:
                print("No more pages available")
                break
                
            # Clear params as they're included in the next_url
            params = {}

        print(f"Total PRs found: {len(all_prs)}")
        if not all_prs:
            print(f"No PRs found for milestone {version}.")
            return False

        # Format PRs into changelog format
        changelog_rows = format_prs_as_changelog(all_prs)
        save_changelog(version, changelog_rows)
        return True

    except requests.exceptions.RequestException as e:
        print(f"Error fetching PRs from GitHub: {e}")
        return False

def extract_changes_section(description: str) -> str:
    """Extract the 'Changes proposed' section from PR description, stripping HTML tags and stopping at the testing section."""
    if not description:
        return ""
        
    # First strip any HTML tags
    description = re.sub(r'<[^>]+>', '', description)
    # Unescape any HTML entities
    description = unescape(description)
    
    # Find content between "Changes proposed" and "How to test"
    pattern = r'Changes proposed in this Pull Request:(.*?)(?=How to test the changes in this Pull Request:|\Z)'
    match = re.search(pattern, description, re.DOTALL)
    if match:
        # Get the captured group (content between the markers)
        content = match.group(1)
        # Clean up the content:
        # 1. Remove any leading/trailing whitespace
        # 2. Remove any empty lines
        # 3. Remove any lines that are just comments
        lines = [line.strip() for line in content.split('\n')]
        lines = [line for line in lines if line and not line.startswith('<!--') and not line.endswith('-->')]
        return '\n'.join(lines).strip()
    return ""

def fetch_pr_description(pr_id: str) -> str:
    """Fetch PR description from GitHub API."""
    # Get GitHub token from environment variable
    token = os.getenv('GITHUB_TOKEN')
    if not token:
        raise ValueError("GITHUB_TOKEN environment variable is not set")
    
    headers = {
        'Authorization': f'token {token}',
        'Accept': 'application/vnd.github.v3+json'
    }
    
    url = f'https://api.github.com/repos/woocommerce/woocommerce/pulls/{pr_id}'
    response = requests.get(url, headers=headers)
    
    if response.status_code == 200:
        return extract_changes_section( response.json()['body'] )
    elif response.status_code == 403:
        print(f"Rate limit exceeded. Waiting 60 seconds...")
        time.sleep(60)
        return fetch_pr_description(pr_id)
    else:
        print(f"Error fetching PR {pr_id}: {response.status_code}")
        return ""

def format_prs_as_changelog(prs):
    """Format PRs into a list of dicts for CSV output."""
    changelog_rows = []
    for pr in prs:
        title = pr["title"]
        pr_id = pr["number"]
        url = pr["html_url"]
        labels = [label["name"] for label in pr.get("labels", []) if label["name"].lower() != "plugin: woocommerce"]
        label_str = ", ".join(labels)
        description = fetch_pr_description(pr_id)
        # description = ''
        changelog_rows.append({
            "ID": pr_id,
            "Title": title,
            "Labels": label_str,
            "URL": url,
            "Description": description,
            "Ranking": ""  # Blank for now
        })
    return changelog_rows

def save_changelog(version, content):
    """Save the changelog content to a CSV file."""
    os.makedirs("changelogs", exist_ok=True)
    filename = f"changelogs/{version}.csv"
    
    # If content is a string (from changelog.txt), convert it to a single row
    if isinstance(content, str):
        rows = [{
            "ID": "",
            "Title": f"Changelog for version {version}",
            "Labels": "",
            "URL": "",
            "Description": content,
            "Ranking": ""
        }]
    else:
        rows = content  # content is already a list of dicts from PRs
    
    fieldnames = ["ID", "Title", "Labels", "URL", "Description", "Ranking"]
    try:
        with open(filename, "w", newline="", encoding="utf-8") as f:
            writer = csv.DictWriter(f, fieldnames=fieldnames)
            writer.writeheader()
            writer.writerows(rows)
        print(f"Successfully saved changelog to {filename}")
    except Exception as e:
        print(f"Error saving changelog: {e}")
        return False
    return True

if __name__ == "__main__":
    # If no version provided as argument, ask for it
    version = sys.argv[1] if len(sys.argv) > 1 else input("Enter WooCommerce version (e.g. 9.8.0): ")
    # success = fetch_changelog(version)
    success = fetch_prs_from_github(version)
    exit(0 if success else 1)

File: fetch-changelog.php
#!/usr/bin/env php
<?php
/**
 * Fetch WooCommerce changelog from GitHub.
 *
 * Usage: php fetch-changelog.php <version>
 *
 * This script fetches the changelog for a specific WooCommerce version,
 * either from the trunk changelog.txt or by querying the GitHub API
 * for PRs in the corresponding milestone.
 */

require_once __DIR__ . '/includes/functions.php';

// Load environment variables
$env = parse_env_file(__DIR__ . '/.env');
$GITHUB_TOKEN = $env['GITHUB_TOKEN'] ?? '';
$REPO_OWNER = 'woocommerce';
$REPO_NAME = 'woocommerce';

/**
 * Fetch changelog from trunk changelog.txt
 *
 * @param string $version The version to fetch
 * @return bool True on success, false on failure
 */
function fetch_changelog($version) {
    $changelog_url = 'https://raw.githubusercontent.com/woocommerce/woocommerce/refs/heads/trunk/changelog.txt';

    $ch = curl_init($changelog_url);
    curl_setopt_array($ch, [
        CURLOPT_RETURNTRANSFER => true,
        CURLOPT_FOLLOWLOCATION => true,
        CURLOPT_HTTPHEADER => ['User-Agent: WooDevTools']
    ]);

    $content = curl_exec($ch);
    $http_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
    curl_close($ch);

    if ($http_code !== 200 || empty($content)) {
        echo "Error fetching changelog from trunk\n";
        return fetch_prs_from_github($version);
    }

    // Find the specific version section
    $version_marker = "= {$version} ";
    if (strpos($content, $version_marker) === false) {
        echo "Changelog section for version {$version} not found. Attempting to generate from GitHub API...\n";
        return fetch_prs_from_github($version);
    }

    // Split at the version marker and get the content up to the next version
    $parts = explode($version_marker, $content);
    if (count($parts) < 2) {
        echo "Error: Could not find changelog section for version {$version}\n";
        return false;
    }

    // Get the content up to the next version marker
    $changelog_parts = explode('= ', $parts[1]);
    $changelog_content = trim($changelog_parts[0]);

    // Save the changelog to a file
    save_changelog($version, $changelog_content);
    return true;
}

/**
 * Fetch PRs from GitHub API for the specified milestone.
 *
 * @param string $version The version to fetch
 * @return bool True on success, false on failure
 */
function fetch_prs_from_github($version) {
    global $GITHUB_TOKEN, $REPO_OWNER, $REPO_NAME;

    if (empty($GITHUB_TOKEN)) {
        echo "Error: GITHUB_TOKEN environment variable is not set\n";
        return false;
    }

    // First verify the milestone exists
    $milestone_url = "https://api.github.com/repos/{$REPO_OWNER}/{$REPO_NAME}/milestones";

    $response = github_request($milestone_url, $GITHUB_TOKEN, true);
    if ($response['http_code'] === 403) {
        echo "Error: GitHub API rate limit exceeded. Please try again later.\n";
        return false;
    }

    $milestones = json_decode($response['body'], true);
    if (!is_array($milestones)) {
        echo "Error: Could not fetch milestones\n";
        return false;
    }

    // Find the milestone that matches our version
    $milestone_number = null;
    foreach ($milestones as $milestone) {
        if ($milestone['title'] === $version) {
            $milestone_number = $milestone['number'];
            echo "Found milestone {$version} with number {$milestone_number}\n";
            break;
        }
    }

    if ($milestone_number === null) {
        echo "Warning: Could not find milestone for version {$version}\n";
        echo "Available milestones:\n";
        foreach ($milestones as $m) {
            echo "- {$m['title']}\n";
        }
        return false;
    }

    // Now fetch PRs with the correct milestone number
    $url = "https://api.github.com/repos/{$REPO_OWNER}/{$REPO_NAME}/issues";
    $params = [
        'state' => 'closed',
        'milestone' => (string)$milestone_number,
        'per_page' => 100
    ];

    $all_prs = [];
    $next_url = $url . '?' . http_build_query($params);
    $page_count = 0;

    while ($next_url) {
        $page_count++;
        echo "Fetching PRs from page {$page_count}...\n";

        $response = github_request($next_url, $GITHUB_TOKEN, true);
        $issues = json_decode($response['body'], true);

        if (!is_array($issues) || empty($issues)) {
            echo "No more PRs found, stopping pagination.\n";
            break;
        }

        // Filter to only include PRs (issues with pull_request field)
        $prs = array_filter($issues, function($issue) {
            return isset($issue['pull_request']);
        });

        echo "Found " . count($prs) . " PRs on this page\n";
        $all_prs = array_merge($all_prs, $prs);

        // Get the next page URL from the Link header
        $next_url = parse_next_page_url($response['headers']['link'] ?? '');

        if (!$next_url) {
            echo "No more pages available\n";
            break;
        }
    }

    echo "Total PRs found: " . count($all_prs) . "\n";

    if (empty($all_prs)) {
        echo "No PRs found for milestone {$version}.\n";
        return false;
    }

    // Format PRs into changelog format
    $changelog_rows = format_prs_as_changelog($all_prs);
    save_changelog($version, $changelog_rows);
    return true;
}

/**
 * Fetch PR description from GitHub API.
 *
 * @param int $pr_id The PR ID
 * @return string The extracted changes section
 */
function fetch_pr_description($pr_id) {
    global $GITHUB_TOKEN, $REPO_OWNER, $REPO_NAME;

    if (empty($GITHUB_TOKEN)) {
        echo "Error: GITHUB_TOKEN environment variable is not set\n";
        return '';
    }

    $url = "https://api.github.com/repos/{$REPO_OWNER}/{$REPO_NAME}/pulls/{$pr_id}";
    $response = github_request($url, $GITHUB_TOKEN, true);

    if ($response['http_code'] === 200) {
        $pr_data = json_decode($response['body'], true);
        return extract_changes_section($pr_data['body'] ?? '');
    } elseif ($response['http_code'] === 403) {
        // Rate limit handled inside github_request, but just in case
        echo "Rate limit exceeded. Waiting 60 seconds...\n";
        sleep(60);
        return fetch_pr_description($pr_id);
    } else {
        echo "Error fetching PR {$pr_id}: HTTP {$response['http_code']}\n";
        return '';
    }
}

/**
 * Format PRs into a list of associative arrays for CSV output.
 *
 * @param array $prs The PRs to format
 * @return array The formatted changelog rows
 */
function format_prs_as_changelog($prs) {
    $changelog_rows = [];

    foreach ($prs as $pr) {
        $title = $pr['title'];
        $pr_id = $pr['number'];
        $url = $pr['html_url'];
        $author = $pr['user']['login'] ?? '';

        // Get labels, excluding "plugin: woocommerce"
        $labels = array_filter($pr['labels'] ?? [], function($label) {
            return strtolower($label['name']) !== 'plugin: woocommerce';
        });
        $label_names = array_map(function($label) {
            return $label['name'];
        }, $labels);
        $label_str = implode(', ', $label_names);

        // Fetch PR description
        $description = fetch_pr_description($pr_id);

        $changelog_rows[] = [
            'ID' => $pr_id,
            'Title' => $title,
            'Author' => $author,
            'Labels' => $label_str,
            'URL' => $url,
            'Description' => $description,
            'Ranking' => '' // Blank for now
        ];
    }

    return $changelog_rows;
}

/**
 * Save the changelog content to a CSV file.
 *
 * @param string $version The version
 * @param mixed $content The content (string or array of rows)
 * @return bool True on success, false on failure
 */
function save_changelog($version, $content) {
    ensure_directory('changelogs');
    $filename = "changelogs/{$version}.csv";

    // If content is a string (from changelog.txt), convert it to a single row
    if (is_string($content)) {
        $rows = [[
            'ID' => '',
            'Title' => "Changelog for version {$version}",
            'Author' => '',
            'Labels' => '',
            'URL' => '',
            'Description' => $content,
            'Ranking' => ''
        ]];
    } else {
        $rows = $content; // content is already a list of dicts from PRs
    }

    $headers = ['ID', 'Title', 'Author', 'Labels', 'URL', 'Description', 'Ranking'];

    if (write_csv_with_bom($filename, $headers, $rows)) {
        echo "Successfully saved changelog to {$filename}\n";
        return true;
    } else {
        echo "Error saving changelog\n";
        return false;
    }
}

// Main execution
if (php_sapi_name() === 'cli') {
    // Get version from argument or prompt
    $version = $argv[1] ?? null;

    if (!$version) {
        echo "Enter WooCommerce version (e.g. 9.8.0): ";
        $version = trim(fgets(STDIN));
    }

    if (empty($version)) {
        echo "Error: Version is required\n";
        echo "Usage: php fetch-changelog.php <version>\n";
        exit(1);
    }

    // Use fetch_prs_from_github directly like the Python script does
    $success = fetch_prs_from_github($version);
    exit($success ? 0 : 1);
}
